File: aps/apslib.rb
# PANDA -- a simple transaction monitor
# 
# Copyright (C) 1998-2005 Ogochan.
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA

#
#	Ruby interface for Exec handler
#

DEBUG=false;

class PandaCore
  def decode(string)
	if  string
	  string.tr('+', ' ').gsub(/((?:%[0-9a-fA-F]{2})+)/n) do
		[$1.delete('%')].pack('H*')
	  end
	else
	  ""
	end
  end
  def encode(val)
	if  val
	  if val.class == String
		string = val;
	  else
		string = val.to_s;
	  end
	  string.gsub(/([^ a-zA-Z0-9_.-]+)/n) do
		'%' + $1.unpack('H2' * $1.size).join('%').upcase
	  end.tr(' ', '+')
	else
	  ""
	end
  end
  def unPack
	rc = 0;
	@values = Hash.new;
	@line.split("&").each { | elem |
	  e = elem.split("=");
	  if e[0] =~ /^dbctrl/
		if e[0] == "dbctrl.rc"
		  rc = decode(e[1]).to_i;
		end
	  else
		@values[e[0]] = decode(e[1]);
	  end
	}
	rc;
  end
  def pack
	str = "";
	@values.each { | key, value |
	  str += key + "=" + encode(value) + "&";
	}
	str.chop;
  end
  def dump(port, name = /.*/)
	@values.each { | key, value |
	  if  name =~ key
		port.printf("%s = [%s]\n",key,value);
	  end
	}
  end
  def []=(name,value)
	if @values[name]
	  @values[name] = value;
	else
	  $stderr.printf("[%s] undefined APS variable\n",name);
	end
  end
  def [](name)
	if @values[name]
	  @values[name];
	else
	  $stderr.printf("[%s] undefined APS variable\n",name);
	  "";
	end
  end
  def clear(rec)
	@values.each { | key, value |
	  if key =~ rec
		@values[key] = "";
	  end
	}
  end
end

class PandaDB < PandaCore
  def dbComm(str)
	if DEBUG
	  $stderr.printf("<<%s\n",str);
	end
	@fpDBW.printf("%s\n",str);
	@fpDBW.flush;
	line = @fpDBR.gets.chop;
	if DEBUG
	  $stderr.printf(">>%s\n",line);
	end
	line;
  end
  def initialize
	@fpDBR = IO.new(3,"r");
	@fpDBW = IO.new(4,"w");
  end
  def execFunction(func)
	str  = "dbctrl.rc=0&";
	str += "dbctrl.func=" + func;
	@line = dbComm(str);
	unPack;
  end
  def fpDBR
	@fpDBR
  end
  def fpDBW
	@fpDBW
  end
end

class PandaTable < PandaDB
  def initialize(db,name)
	@name = name;
	@db = db;
	@values = Hash.new;
	str  = "dbctrl.rc=0&";
	str += "dbctrl.rname=" + @name;
	@line = @db.dbComm(str);
	unPack;
  end
  def getSchema(db,pname = "", func = "")
	@values = Hash.new;
	str  = "dbctrl.rc=%2D1&";
	str += "dbctrl.func=" + func + "&";
	str += "dbctrl.rname=" + @name + "&";
	str += "dbctrl.pname=" + pname;
	@line = @db.dbComm(str);
	unPack;
  end
  def execFunction(func, pname = "")
	str  = "dbctrl.rc=0&";
	str += "dbctrl.func=" + func + "&";
	str += "dbctrl.rname=" + @name + "&";
	str += "dbctrl.pname=" + pname + "&";
	str += pack;
	@line = @db.dbComm(str);
	unPack;
  end
end

class PandaDC < PandaCore
  def initialize
	line = $stdin.gets
	@line = line.chomp;
	@db = PandaDB.new;
	unPack;
  end
  def run
	case @values["mcparea.dc.status"]
	when "LINK"
	  start(self, @db);
	when "PUTG"
	  name = "do_" + @values["mcparea.dc.event"];
	  if respond_to?(name)
		__send__(name, self, @db);
	  end
	else
	end
	out;
  end
  def out
	str = pack;
	printf("%s\n",str);
	$stdout.flush;
  end
  def putWindow(name,widget = "" ,type = "")
	@values["mcparea.dc.window"] = name;
	@values["mcparea.dc.widget"] = widget;
	@values["mcparea.dc.puttype"] = type;
	@values["mcparea.dc.status"] = "PUTG";
	@values["mcparea.rc"] = 0;
  end
  WidgetNormal = 0;
  WidgetActive = 1;
  WidgetPreLight = 2;
  WidgetSelected = 3
  WidgetInSensitive = 4;
end

class PandaBatch < PandaCore
  def initialize
	@db = PandaDB.new;
  end
  def run
	start(@db);
  end
end

File: aps/apslib.c
/*
 * PANDA -- a simple transaction monitor
 * Copyright (C) 2000-2008 Ogochan.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

/*
#define	DEBUG
#define	TRACE
*/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <glib.h>

#include "const.h"
#include "libmondai.h"
#include "directory.h"
#include "handler.h"
#include "dbgroup.h"
#include "apslib.h"
#include "debug.h"

extern int MCP_PutWindow(ProcessNode *node, char *wname, char *type) {
  ValueStruct *mcp;

  mcp = node->mcprec->value;
  if (wname != NULL) {
    dbgprintf("window = [%s]", wname);
    SetValueString(GetItemLongName(mcp, "dc.window"), wname, NULL);
  }
  SetValueString(GetItemLongName(mcp, "dc.widget"), "", NULL);
  SetValueString(GetItemLongName(mcp, "dc.puttype"), type, NULL);
  SetValueString(GetItemLongName(mcp, "dc.status"), "PUTG", NULL);
  SetValueInteger(GetItemLongName(mcp, "rc"), 0);
  return (0);
}

extern int MCP_ReturnComponent(ProcessNode *node, char *event) {
  ValueStruct *mcp;

  mcp = node->mcprec->value;
  if (event != NULL) {
    SetValueString(GetItemLongName(mcp, "dc.event"), event, NULL);
  }
  SetValueString(GetItemLongName(mcp, "dc.puttype"), MCP_PUT_RETURN, NULL);
  SetValueString(GetItemLongName(mcp, "dc.status"), "PUTG", NULL);
  SetValueInteger(GetItemLongName(mcp, "rc"), 0);
  return (0);
}

extern RecordStruct *MCP_GetWindowRecord(ProcessNode *node, char *name) {
  WindowBind *bind;
  RecordStruct *ret;

  bind = (WindowBind *)g_hash_table_lookup(node->bhash, name);
  ret = bind->rec;
  return (ret);
}

extern void *MCP_GetEventHandler(GHashTable *StatusTable, ProcessNode *node) {
  char *status, *event;
  GHashTable *EventTable;
  void (*handler)(ProcessNode *);

  status =
      ValueStringPointer(GetItemLongName(node->mcprec->value, "dc.status"));
  event = ValueStringPointer(GetItemLongName(node->mcprec->value, "dc.event"));

  dbgprintf("status = [%s]", status);
  dbgprintf("event  = [%s]", status, event);
  if ((EventTable = g_hash_table_lookup(StatusTable, status)) == NULL) {
    EventTable = g_hash_table_lookup(StatusTable, "");
  }
  if (EventTable != NULL) {
    if ((handler = g_hash_table_lookup(EventTable, event)) == NULL) {
      handler = g_hash_table_lookup(EventTable, "");
    }
  } else {
    handler = NULL;
  }

  return (handler);
}

extern void MCP_RegistHandler(GHashTable *StatusTable, char *status,
                              char *event, void (*handler)(ProcessNode *)) {
  GHashTable *EventTable;

  if ((EventTable = g_hash_table_lookup(StatusTable, status)) == NULL) {
    EventTable = NewNameHash();
    g_hash_table_insert(StatusTable, StrDup(status), EventTable);
  }
  if (g_hash_table_lookup(EventTable, event) == NULL) {
    g_hash_table_insert(EventTable, StrDup(event), handler);
  }
}

extern ValueStruct *MCP_GetDB_Define(char *name) {
  char buff[SIZE_LONGNAME + 1];
  int rno, pno, ono;
  RecordStruct *rec;
  PathStruct *path;
  DB_Operation *op;
  ValueStruct *val, *ret;
  char *p, *pname, *oname;

  strcpy(buff, name);
  if ((p = strchr(buff, ':')) != NULL) {
    *p = 0;
    pname = p + 1;
    if ((p = strchr(pname, ':')) != NULL) {
      *p = 0;
      oname = p + 1;
    } else {
      oname = NULL;
    }
  } else {
    oname = NULL;
    pname = NULL;
  }

  val = NULL;
  if ((rno = (int)(long)g_hash_table_lookup(DB_Table, name)) != 0) {
    rec = ThisDB[rno - 1];
    val = rec->value;
    if ((pname != NULL) && ((pno = (int)(long)g_hash_table_lookup(
                                 rec->opt.db->paths, pname)) != 0)) {
      path = rec->opt.db->path[pno - 1];
      val = (path->args != NULL) ? path->args : val;
      if ((oname != NULL) &&
          ((ono = (int)(long)g_hash_table_lookup(path->opHash, oname)) != 0)) {
        op = path->ops[ono - 1];
        val = (op->args != NULL) ? op->args : val;
      }
    }
  }
  if (val != NULL) {
    ret = DuplicateValue(val, FALSE);
  } else {
    ret = NULL;
  }
  return (ret);
}
